distributed/tests/test_worker_state_machine.py

```diff
@@ -302,3 +302,37 @@ async def test_fetch_via_amm_to_compute(c, s, a, b):
     b.comm_threshold_bytes = old_comm_threshold
 
     await f1
+
+
+@gen_cluster(client=True)
+async def test_cancelled_while_in_flight(c, s, a, b):
+    event = asyncio.Event()
+    a.rpc = _LockedCommPool(a.rpc, write_event=event)
+
+    x = c.submit(inc, 1, key="x", workers=[b.address])
+    y = c.submit(inc, x, key="y", workers=[a.address])
+    await wait_for_state("x", "flight", a)
+    y.release()
+    await wait_for_state("x", "cancelled", a)
+
+    # Let the comm from b to a return the result
+    event.set()
+    # upon reception, x transitions cancelled->forgotten
+    while a.tasks:
+        await asyncio.sleep(0.01)
+
+
+@gen_cluster(client=True)
+async def test_in_memory_while_in_flight(c, s, a, b):
+    event = asyncio.Event()
+    a.rpc = _LockedCommPool(a.rpc, write_event=event)
+
+    x = c.submit(inc, 1, key="x", workers=[b.address])
+    y = c.submit(inc, x, key="y", workers=[a.address])
+    await wait_for_state("x", "flight", a)
+    a.update_data({"x": 3})
```

Comment:
if x is already in flight, how could the data end up in memory without us doing it explicitly like in this unit test?
acquire_replica and "fetch_dependency" should not fetch this key a second time.

From me reading the code, the only way this could happen is via `Client.scatter`. I would argue a user should not be allowed to scatter a key that is already known to the cluster to be computed.

I don't want to block this PR for this test but if the above outline is how we end up in this situation, I think we should prohibit scattering such keys and shrink the space of possible/allowed transitions.

Specifically, I'm inclined to say `a.update({"x": 3})` should raise an exception if x is in flight. 

Thoughts?

Reply:
> Specifically, I'm inclined to say a.update({"x": 3}) should raise an exception if x is in flight.

would might translate to something like

```python
def transition_flight_memory(...):
    if not ts.done:
        raise ImpossibleTransition("A nice exception that tells us that we cannot move data to memory while in flight but coro/task still running")
```

(Where the exception is supposed to be raised is not the point of my argument. It may not be feasible to raise *in* the transition itself, idk)

Reply:
The more I think about this the stronger I feel about it because these kind of race conditions are part of why I introduced cancelled/resumed to avoid us needing to deal with these transitions. 
If the fetch task would finish successfully, this would cause a memory->memory transition. Since this is not allowed/possible this would cause a

memory->released
(possibly the released transition would cancel some follow up tasks)
released->memory

or as a concrete story

```python
[
    (ts.key, "flight", "memory", "memory", {dependent: "executing"})
    (dependent.key, "waiting", "executing", "executing", {}),
    # A bit later after gather_dep returns
    (ts.key, "memory", "memory", "released", {dependent: "released"}),
    (dependent.key, "executing", "released", "cancelled", {}),
    (ts.key, "released", "memory", "memory", {dependent: "waiting"}),
    (dependent.key, "cancelled", "waiting", "executing", {}),
]
```

writing down the expected story made me realize that our transition flow _should_ heal us here but we'd be performing a lot of unnecessary transitions that could expose us to problems.

Reply:
I think that 80% of the problem is caused by the non-sequentiality of RPC calls vs. bulk comms.

1. client scatters to a
2. the scheduler does not know about scattered keys until the three-way round-trip between client, workers, and scheduler has been completed:
https://github.com/dask/distributed/blob/fb3589c9596b7eee25db0655fd24694ed2035371/distributed/scheduler.py#L5018-L5022
3. in the middle of that handshake, a client (not necessarily the same client) calls compute on b and then gather_dep to copy the key from b to a
4. while the flight from b to a is in progress, the scatter finishes, which triggers update_data as shown in the test.

The only way to avoid this would be to fundamentally rewrite the scatter implementation. Which, for the record, I think is long overdue.

Reply:
I'll explain the above in a comment in the test

Reply:
My point is only partially about technical correctness of race conditions but also about whether this is even a sane operation. How can a user know the value of `x` if `x` is supposed to be computed on the cluster?